### PR TITLE
Implement user authentication

### LINKED ISSUE #55: Implementing user authentication
**Is your feature request related to a problem? Please describe.**
The current state of the application lacks basic authentication for the frontend, making it necessary to implement this essential functionality.

**Describe the solution you'd like**
This issue aims to address the following functionalities to enhance the application's security and user experience:

User Authentication: Implement a login mechanism that allows users to log in using their credentials (email and password) while utilizing HTTP-only Cookies for enhanced security.

User Data Storage: Upon successful authentication, save the authenticated user's data in context. This context should be accessible from any part of the application, ensuring that user-specific information is available where needed.

By implementing these solutions, we will establish a foundation for user authentication and data management, enhancing the overall functionality and security of the application.

**Describe alternatives you've considered**

**Additional context**

### PR DESCRIPTION
**Pull request by Madhawa97**
## Purpose

The purpose of this PR is to fix #55 

## Goals

Implement a robust and user-friendly login mechanism that allows users to securely log in using their email and password credentials.

## Approach

Implement a login mechanism that allows users to log in using their credentials (email and password) while utilizing HTTP-only Cookies for enhanced security.
Upon successful authentication, save the authenticated user's data in context. This context should be accessible from any part of the application, ensuring that user-specific information is available where needed.

## Screenshots
image

##  Checklist
- [x] This PR doesn't commit any keys, passwords, tokens, usernames, or other secrets.
- [x] I have read and understood the development best practices guidelines ( http://bit.ly/sef-best-practices )
- [x] My code follows the style guidelines of this project
- [x] I have performed a self-review of my own code
- [x] I have commented my code, particularly in hard-to-understand areas
- [ ] I have made corresponding changes to the documentation

## Test environment
 
Node v18, macOS latest, chrome latest

## Learning

User Authentication, Typescript, useContext

### CODE REVIEW

File: src/components/Layout/Navbar/Navbar.tsx [hunk 1]
```diff
@@ -10,11 +10,13 @@ import { Button, Col, Row, Space, Typography } from 'antd';
 
 import styles from './Navbar.module.css';
 import MenuDrawer from '../MenuDrawer/MenuDrawer';
+import JoinUsDialog from '../../JoinUsDialog/JoinUsDialog';
 
 const { Text } = Typography;
 
 const Navbar: React.FC = () => {
   const [openMenu, setOpenMenu] = useState(false);
+  const [isJoinUsDialogOpen, setJoinUsDialogOpen] = useState(false);
```
Comment (anjula-sack) [inline src/components/Layout/Navbar/Navbar.tsx]:
```suggestion
  const [isLoginModalVisible, setIsLoginModalVisible] = useState(false);
```

File: src/components/Layout/Navbar/Navbar.tsx [hunk 2]
```diff
@@ -10,11 +10,13 @@ import { Button, Col, Row, Space, Typography } from 'antd';
 
 import styles from './Navbar.module.css';
 import MenuDrawer from '../MenuDrawer/MenuDrawer';
+import JoinUsDialog from '../../JoinUsDialog/JoinUsDialog';
```
Comment (anjula-sack) [inline src/components/Layout/Navbar/Navbar.tsx]:
```suggestion
import LoginModal from '../../LoginModal';
```

File: src/contexts/UserContext.tsx [hunk 3]
```diff
@@ -0,0 +1,46 @@
+import React, { createContext, useState, useEffect, useContext } from 'react';
+import axios, { type AxiosResponse } from 'axios';
+import type { Profile } from '../types';
+import { API_URL } from './../constants';
```
Comment (anjula-sack) [inline src/contexts/UserContext.tsx:4]:
```suggestion
import { API_URL } from './../constants';

```

File: src/contexts/UserContext.tsx [hunk 4]
```diff
@@ -0,0 +1,46 @@
...
+export interface UserContextType {
+  user: any;
+  setUser: any;
```
Comment (anjula-sack) [inline src/contexts/UserContext.tsx]: Add the type

File: src/components/JoinUsDialog/JoinUsDialog.tsx [hunk 5]
```diff
@@ -0,0 +1,173 @@
+import React, { useState, useContext } from 'react';
+import axios, { type AxiosResponse } from 'axios';
+import { API_URL } from '../../constants';
+import { type Profile } from '../../types';
+import UserContext from '../../contexts/UserContext';
+
+const JoinUsDialog: React.FC<{ isOpen: boolean; onClose: () => void }> = ({
```
Comment (anjula-sack) [inline src/components/JoinUsDialog/JoinUsDialog.tsx]: interface LoginModalProps {

}

File: src/components/Layout/Navbar/Navbar.tsx [hunk 6]
```diff
@@ -89,11 +91,24 @@ const Navbar: React.FC = () => {
...
+            </Button>
           </Space>
         </Col>
       </Row>
       <MenuDrawer openMenu={openMenu} setOpenMenu={setOpenMenu} />
+      <JoinUsDialog
+        isOpen={isJoinUsDialogOpen}
+        onClose={() => {
+          setJoinUsDialogOpen(false);
+        }}
```
Comment (anjula-sack) [inline src/components/Layout/Navbar/Navbar.tsx]:
```suggestion
        onClose={setJoinUsDialogOpen(false)}
```

File: src/components/JoinUsDialog/JoinUsDialog.tsx [hunk 7]
```diff
@@ -0,0 +1,173 @@
...
+              viewBox="0 0 24 24"
+              stroke="currentColor"
+            >
+              <path
+                strokeLinecap="round"
+                strokeLinejoin="round"
+                strokeWidth="2"
+                d="M6 18L18 6M6 6l12 12"
+              />
+            </svg>
```
Comment (anjula-sack) [inline src/components/JoinUsDialog/JoinUsDialog.tsx]: Move svgs to assets

File: src/components/Layout/Navbar/Navbar.tsx (hunk 6 above)
Comment (anjula-sack) [reply src/components/Layout/Navbar/Navbar.tsx]: {isLoginModalVisible ? <LoginModal> : null}

File: src/App.tsx [hunk 8]
```diff
@@ -1,12 +1,19 @@
...
 
-const App: React.FC = () => (
-  <MainLayout>
-    <Home />
-  </MainLayout>
-);
+const App: React.FC = () => {
+  const { user, setUser } = useContext(UserContext);
+
+  user !== null && console.log('user is authenticated');
```
Comment (anjula-sack) [inline src/App.tsx]: Move this into useEffect

**Review by anjula-sack — CHANGES_REQUESTED** [review]

File: src/components/Layout/Navbar/Navbar.tsx (hunk 2 above)
Comment (Madhawa97) [reply src/components/Layout/Navbar/Navbar.tsx]: Done

File: src/components/Layout/Navbar/Navbar.tsx (hunk 1 above)
Comment (Madhawa97) [reply src/components/Layout/Navbar/Navbar.tsx]: Done

File: src/contexts/UserContext.tsx (hunk 4 above)
Comment (Madhawa97) [reply src/contexts/UserContext.tsx]: Done

File: src/components/JoinUsDialog/JoinUsDialog.tsx (hunk 5 above)
Comment (Madhawa97) [reply src/components/JoinUsDialog/JoinUsDialog.tsx]: Done

File: src/App.tsx (hunk 8 above)
Comment (Madhawa97) [reply src/App.tsx]: the `useEffect` was defined in the `UserProvider` to invoke the `getUser()` each time the `UserProvider` mounts.
But as requested I changed it to invoke the `getUser()` each time the `App` mounts. 
and the `getUser()` has also been sent to the `App` via Context. Could you let me know which way is correct/preferred?

File: src/components/Layout/Navbar/Navbar.tsx [hunk 9]
```diff
@@ -89,11 +99,20 @@ const Navbar: React.FC = () => {
                 <InstagramOutlined className={styles.antIcon} />
               </Button>
             </a>
-            <Button className={styles.loginButton}>Join Us</Button>
+            <Button
+              className={styles.loginButton}
+              onClick={handleLoginModalOpen}
+            >
+              Join Us
```
Comment (anjula-sack) [inline src/components/Layout/Navbar/Navbar.tsx]:
```suggestion
              Login
```

File: src/components/Layout/Navbar/Navbar.tsx [hunk 10]
```diff
@@ -89,11 +99,20 @@ const Navbar: React.FC = () => {
...
+            </Button>
           </Space>
         </Col>
       </Row>
       <MenuDrawer openMenu={openMenu} setOpenMenu={setOpenMenu} />
+      <LoginModal
+        isLoginModalVisible={isLoginModalVisible}
+        onClose={handleLoginModalClose}
+      />
```
Comment (anjula-sack) [inline src/components/Layout/Navbar/Navbar.tsx]:
```suggestion
  {  isLoginModalVisible ?  <LoginModal
        isLoginModalVisible={isLoginModalVisible}
        onClose={handleLoginModalClose}
      /> : null }
```

File: src/components/LoginModal/index.tsx [hunk 11]
```diff
@@ -0,0 +1,163 @@
...
+          } else {
+            setUserContext(null);
+          }
+        });
+    } catch (error) {
+      console.error('Login error:', error);
+    }
+  };
+
+  return isLoginModalVisible ? (
```
Comment (anjula-sack) [inline src/components/LoginModal/index.tsx]:
```suggestion
  return (
```

File: src/components/LoginModal/index.tsx [hunk 12]
```diff
@@ -0,0 +1,163 @@
...
+              <div className="text-sm font-medium text-gray-900">
+                Not registered yet?{' '}
+                <a className="text-blue-600 hover:underline">Create account</a>
+              </div>
+            </form>
+          </div>
+        </div>
+      </div>
+    </div>
+  ) : null;
```
Comment (anjula-sack) [inline src/components/LoginModal/index.tsx]:
```suggestion
  );
```

**Review by anjula-sack — CHANGES_REQUESTED** [review]

**Review by anjula-sack — APPROVED** [review]
Thank you! @Madhawa97

**Pull request merged**